File: modules/ssh-module/ios/SshModule.swift
import ExpoModulesCore

struct TransferProgressSwift: Record {
  @Field
  var transferred: String
  @Field
  var total: String
}

struct ConnectionDetails {
    var user: String
    var password: String?
    var key: String?
    var address: String
}

public class SshModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  var session: Session?
    
  var connectionDetails: ConnectionDetails?
    
    func reconnect() async throws -> Session {
        guard let connectionDetails = self.connectionDetails else {
            throw NSError(domain: "app.reflect.serverremote", code: 1, userInfo: [NSLocalizedDescriptionKey: "Session is null"])
        }
        guard let session = self.session else {
            try await connection(user: connectionDetails.user, password: connectionDetails.password, key: connectionDetails.key, addrs: connectionDetails.address)
            throw NSError(domain: "app.reflect.serverremote", code: 1, userInfo: [NSLocalizedDescriptionKey: "unreachable"])
        }
        do {
            _ = try await session.testConnection()
        } catch {
            print("test_connection error: "+String(describing: error))
            try await connection(user: connectionDetails.user, password: connectionDetails.password, key: connectionDetails.key, addrs: connectionDetails.address)
        }
        return self.session!
    }
    
    func connection(user: String, password: String?, key: String?, addrs: String) async throws -> Void {
        guard password?.isEmpty == false || key?.isEmpty == false else {
            throw NSError(domain: "app.reflect.serverremote", code: 1, userInfo: [NSLocalizedDescriptionKey: "Please input password or key"])
        }
        if password?.isEmpty == false {
            try await connectionPassword(user: user, password: password!, addrs: addrs)
        } else {
            try await connectionKey(user: user, password: password, key: key!, addrs: addrs)
        }

    }
    
    func connectionPassword(user: String, password: String, addrs: String) async throws -> Void {
        do {
            self.session = try connect(user: user, password: password, addrs: addrs)
            self.connectionDetails = ConnectionDetails(user: user, password: password, address: addrs)
        } catch {
            print("connect error: "+String(describing: error))
            throw error
        }
    }

    func connectionKey(user: String, password: String?, key: String, addrs: String) async throws -> Void {
        do {
            self.session = try connectKey(user: user, key: key, password: password, addrs: addrs)
            self.connectionDetails = ConnectionDetails(user: user, password: password, address: addrs)
        } catch {
            print("connect error: "+String(describing: error))
            throw error
        }
    }

  
  public func definition() -> ModuleDefinition {

    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('SshModule')` in JavaScript.
    Name("SshModule")

    // Sets constant properties on the module. Can take a dictionary or a closure that returns a dictionary.
    Constants([
      "PI": Double.pi
    ])

    // Defines event names that the module can send to JavaScript.
    Events("exec")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      return "Hello world! 👋"
    }

    AsyncFunction("connect") { (user: String, password: String, key: String?, addrs: String) async throws -> Void in
        try await connection(user: user, password: password, key: key, addrs: addrs)
    }

    AsyncFunction("exec") { (commandId: String, command: String) async throws -> String in

        let session = try await self.reconnect()
        
        let isCompleted = IsCompleted()

        @Sendable func getData() async throws {
            do {
                while let data = try await session.readOutput(commandId: commandId) {
                    print("exec data: "+String(describing: data))
                    self.sendEvent("exec", ["commandId": commandId, "data": data])
                }
            } catch {
                print("read output error: "+String(describing: error))
                throw error
            }
        }

        let task = Task {
            while true {
                let isCompleted = await isCompleted.get()
                if isCompleted {
                    break
                }
                try await Task.sleep(nanoseconds: 100_000_000) // Sleep for 100ms
                try await getData()
            }
        }
        
        do {
            let returnCode = try await session.exec(commandId: commandId, command: command)
            await isCompleted.set(true)
            try await task.value
            try await getData()
            self.sendEvent("exec", ["commandId": commandId, "data": "eventingComplete"])
            
            return returnCode
        } catch {
            print("exec error: "+String(describing: error))
            throw error
        }
    }
      
      AsyncFunction("cancel") { (id: String) async throws -> String in
          let session = try await self.reconnect()

          try await session.cancel(id: id)
          return "0"
      }
      
      AsyncFunction("transfer") { (transferId: String, sourcePath: String, destinationPath: String, direction: String) async throws -> String in
          let session = try await self.reconnect()

          do {
              print("trying transfer")
              print(transferId, sourcePath, destinationPath, direction)
              try await session.transfer(transferId: transferId, sourcePath: sourcePath, destinationPath: destinationPath, direction: direction)
              print("success")
          } catch {
              print("transfer error: "+String(describing: error))
              throw error
          }
          return "0"
      }
      // TODO: wrap all these functions in an inner function, retry once with a reconnect if there is an error
      AsyncFunction("transferProgress") { (transferId: String) async throws -> TransferProgressSwift in
          let session = try await self.reconnect()

          do {
              let progress = try await session.transferProgress(transferId: transferId)
              return TransferProgressSwift(
                transferred: Field(wrappedValue: String(progress.transferred)),
                total: Field(wrappedValue: String(progress.total))
              )
          } catch {
              print("transfer progress error: "+String(describing: error))
              throw error
          }
      }

  }
}

actor IsCompleted {
    private var isCompleted = false
    func get() -> Bool {
        return isCompleted
    }
    func set(_ value: Bool) {
        isCompleted = value
    }
}

File: modules/ssh-module/android/src/main/java/expo/modules/sshmodule/SshModule.kt
package expo.modules.sshmodule

import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import uniffi.rust_lib.Session
import uniffi.rust_lib.connect
import android.util.Log
import kotlinx.coroutines.*
import uniffi.rust_lib.connectKey
import java.util.concurrent.atomic.AtomicBoolean



class TransferProgress(
  @Field
  val transferred: String,
  @Field
  val total: String
) : Record

class ConnectionDetails(
    val user: String,
    val password: String?,
    val key: String?,
    val address: String
)

class SshModule : Module() {
  private var session: Session? = null
  private var connectionDetails: ConnectionDetails? = null

  private suspend fun reconnect(): Session {
      val details = connectionDetails ?: throw CodedException("Session is null")
      val currentSession = session ?: return connection(details.user, details.password, details.key, details.address)
      try {
          currentSession.testConnection()
          return currentSession
      } catch (e: Exception) {
          return connection(details.user, details.password, details.key, details.address)
      }
  }

  private suspend fun connection(user: String, password: String?, key: String?, address: String): Session {
      if (password.isNullOrEmpty() && key.isNullOrEmpty()) {
          throw CodedException("Please input password or key")
      }

      if (!password.isNullOrEmpty()) {
          return connectionPassword(user, password, address)
      } else {
          return connectionKey(user, password, key!!, address)
      }
  }

  private suspend fun connectionPassword(user: String, password: String, address: String): Session {
      try {
          session = connect(user, password, address)
          connectionDetails = ConnectionDetails(user, password, null, address)
          Log.d("kotlin", "finished connecting with password")
          return session!!
      } catch (e: Exception) {
          Log.d("kotlin", "connect error: $e")
          throw e
      }
  }

  private suspend fun connectionKey(user: String, password: String?, key: String, address: String): Session {
      try {
          session = connectKey(user, key, password, address)
          connectionDetails = ConnectionDetails(user, password, key, address)
          Log.d("kotlin", "finished connecting with key")
          return session!!
      } catch (e: Exception) {
          Log.d("kotlin", "connect error: $e")
          throw e
      }
  }

  override fun definition() = ModuleDefinition {
    Name("SshModule")

    Constants(
      "PI" to Math.PI
    )


    Events("exec")

    Function("hello") {
      "Hello world! 👋"
    }

    AsyncFunction("connect") { user: String, password: String, key: String?, address: String, promise: Promise ->
      Log.d("kotlin", "starting connect")
      CoroutineScope(Dispatchers.IO).launch {
          try {
              connection(user, password, key, address)
              promise.resolve("0")
          } catch (e: Exception) {
              promise.reject(CodedException(e))
          }
      }
    }

    AsyncFunction("exec") { commandId: String, command: String, promise: Promise ->
      println("kotlin: precoroutine")
      CoroutineScope(Dispatchers.IO).launch {
        try {
          Log.d("kotlin", "going to reconnect")
          val session = reconnect()

          // Use an atomic flag to signal when the exec has completed.
          val isCompleted = AtomicBoolean(false)

          // Suspend function to read output and send events.
          suspend fun getData() {
            try {
              // Loop until no more output is available.
              while (true) {
                // Assume readOutput is a suspend function returning a nullable String.
                val data = session.readOutput(commandId) ?: break
                Log.d("kotlin", data)
                sendEvent("exec", mapOf("commandId" to commandId, "data" to data))
              }
            } catch (e: Exception) {
              Log.d("kotlin", "read output error: ${e}")
              throw e
            }
          }

          // Launch a concurrent job that periodically calls getData().
          val job = launch {
            while (!isCompleted.get()) {
              delay(100) // 100ms delay similar to Swift's sleep(nanoseconds: 100_000_000)
              getData()
            }
          }

          Log.d("kotlin", "starting exec")
          // Run the exec command; this is assumed to be a suspend function returning a String.
          val output = session.exec(commandId, command)
          Log.d("kotlin", "exec returned")
          Log.d("kotlin", output)

          // Signal that exec has completed.
          isCompleted.set(true)
          // Wait for the background job to finish.
          job.join()
          // Do one final getData() call to catch any remaining output.
          getData()
          // Send the final event indicating that eventing is complete.
          sendEvent("exec", mapOf("commandId" to commandId, "data" to "eventingComplete"))

          // Resolve the promise with the exec output.
          promise.resolve(output)
        } catch (e: Exception) {
          promise.reject(CodedException(e))
        }
      }
    }


    AsyncFunction("cancel") { id: String, promise: Promise ->
      CoroutineScope(Dispatchers.IO).launch {
        try {
          val session = reconnect()
          session.cancel(id)
          promise.resolve("0")
        } catch (e: Exception) {
          promise.reject(CodedException(e))
        }
      }
    }

    AsyncFunction("transfer") { transferId: String, sourcePath: String, destinationPath: String, direction: String, promise: Promise ->
      CoroutineScope(Dispatchers.IO).launch {
        try {
          val session = reconnect()
          session.transfer(transferId, sourcePath, destinationPath, direction)
          promise.resolve("0")
        } catch (e: Exception) {
          promise.reject(CodedException(e))
        }
      }
    }

    AsyncFunction("transferProgress") { transferId: String, promise: Promise ->
      CoroutineScope(Dispatchers.IO).launch {
        try {
          val session = reconnect()
          val progress = session.transferProgress(transferId)
          promise.resolve(TransferProgress(progress.transferred.toString(), progress.total.toString()))
        } catch (e: Exception) {
          promise.reject(CodedException(e))
        }
      }
    }
  }
}
